### PR TITLE
HIVE-27918: Iceberg: Push transforms for clustering during table writes

### PR DESCRIPTION
**Pull request by SourabhBadhya**
### What changes were proposed in this pull request?

Iceberg: Push transforms for clustering during table writes

### Why are the changes needed?

Currently transformed columns (except for bucket transform) are not pushed / passed as clustering columns. This can lead to incorrect clustering on such columns which can lead non-performant writes. Hence push transforms for clustering during table writes.

### Does this PR introduce _any_ user-facing change?

No

### Is the change a dependency upgrade?

No

### How was this patch tested?

QTests

### CODE REVIEW

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java [hunk 1]
```diff
@@ -234,6 +240,77 @@ public class HiveIcebergStorageHandler implements HiveStoragePredicateHandler, H
             }
           };
 
+  private static final transient BiFunction<Integer, Integer, Function<List<ExprNodeDesc>, ExprNodeDesc>>
```
Comment (deniskuzZ) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java:243]: this class looks overloaded already, could we move these functions into some helper class?

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java [hunk 2]
```diff
@@ -732,6 +809,16 @@ private void addCustomSortExpr(Table table,  org.apache.hadoop.hive.ql.metadata.
       int order = fieldOrderMap.get(spec.getColumnName());
       if (TransformSpec.TransformType.BUCKET.equals(spec.getTransformType())) {
         customSortExprs.add(BUCKET_SORT_EXPR.apply(order + offset, spec.getTransformParam().get()));
+      } else if (TransformSpec.TransformType.TRUNCATE.equals(spec.getTransformType())) {
```
Comment (deniskuzZ) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java:812]: that's enum, could we use switch instead?

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java (hunk 1 above)
Comment (SourabhBadhya) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java]: Done.

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java (hunk 2 above)
Comment (SourabhBadhya) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java]: I have replaced the if-else block with switch block.

**Pull request merged**